### ISSUE TITLE
tornado 5.1.1 is not co-installable with salt on python2.7

### ISSUE THREAD

**Issue by prometheanfire** [saltstack/salt#50403]
### Description of Issue/Question
Salt states that it works with tornado<6.0.0 in base.txt but the metadata it installs does not follow the python version specifier.

    # grep tornado ~/.virtualenvs/salt-tornado/lib/python2.7/site-packages/salt-2018.3.3.dist-info/METADATA
    Requires-Dist: tornado (<6.0,>=4.2.1)
    Requires-Dist: tornado (<5.0,>=4.2.1)

When pip installing you get these lines relating to tornado.
    Ignoring tornado: markers 'python_version >= "3.4"' don't match your environment
    Requirement already satisfied: tornado<6.0,>=4.2.1; python_version < "3" in /home/mthode/.virtualenvs/salt-tornado/lib/python2.7/site-packages (from salt==2018.3.3) (5.1.1)
    salt 2018.3.3 has requirement tornado<5.0,>=4.2.1, but you'll have tornado 5.1.1 which is incompatible.

### Setup
no setup

### Steps to Reproduce Issue
1. Create new virtualenv
2. pip install tornado===5.1.1
3. pip install salt (or pip install .)

### Versions Report
not installed, 2018.3.3 tested (from tag)

**Comment by prometheanfire** [comment]
It seems to me you are doing a ton of custom stuff in setup.py that's shooting yourself in the foot.  Just use setuptools/pkg_resources/etc for the requirements stuff. you can specify playforms, optional requirements (extras), etc.

**Comment by dwoz** [comment]
@prometheanfire Thank you for reporting this. I have confirmed the issue.

**Comment by dwoz** [comment]
@prometheanfire Can you provide us with the output of `salt --versions-report` and if you are unable to run that due to a broken install can you provide us the OS and Python Version?

**Comment by prometheanfire** [comment]
keep in mind this is in a venv, but here it is.

```
Salt Version:
           Salt: 2018.3.0
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.15 (default, Sep 13 2018, 01:26:59)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 5.1.1
            ZMQ: 4.2.5
 
System Versions:
           dist: gentoo 2.4.1 
         locale: UTF-8
        machine: x86_64
        release: 4.18.16-gentoo
         system: Linux
        version: Gentoo Base System 2.4.1 
```

**Comment by prometheanfire** [comment]
I noticed this bug in openstack co-installability tests for proposed updates. jaeger-client has this cap as well iirc.

http://logs.openstack.org/31/615731/1/check/requirements-tox-py27-check-uc/3712d13/job-output.txt.gz#_2018-11-06_06_35_48_249574

**Comment by dwoz** [comment]
Looks like we migrated to pipenv our setup.py isn't able to handle the requirements yet. [This pipenv](https://github.com/pypa/pipenv/issues/209) issue has a recipe for integrating with setup.py

**Comment by gtmanfred** [comment]
I just want to state for the record, salt never migrated to pipenv.  There is a Pipfile for people that wanted it, but all that was added was markers to the requirements.txt files.

**Comment by prometheanfire** [comment]
https://github.com/saltstack/salt/commit/9c6f65363e58913971ec21c3021b789f9d65d7a9#diff-2eeaed663bd0d25b7e608891384b7298

should fix it, testing now

**Comment by prometheanfire** [comment]
https://review.openstack.org/639256 for those curious

**Comment by prometheanfire** [comment]
I do wonder when this will be fixed

https://github.com/saltstack/salt/commit/aeacd2b749774857880821c0586e6fe0f46bafbd

**Comment by gtmanfred** [comment]
Neon https://github.com/saltstack/salt-jenkins/issues/995

**Comment by devurandom** [comment]
> saltstack/salt-jenkins#995

That report was moved to #51883.

**Comment by devurandom** [comment]
@prometheanfire Is it possible to apply the patch you referenced in the Gentoo ebuild?

**Comment by prometheanfire** [comment]
I don't see any patches that do anything for this
(see https://github.com/gentoo/gentoo/tree/master/app-admin/salt/files and https://github.com/gentoo/gentoo/tree/master/app-admin/salt )

**Comment by dwoz** [comment]
This was resolved by #55952 in version `3000`
